mlflow/litellm/__init__.py

```diff
@@ -0,0 +1,144 @@
+from mlflow.utils.autologging_utils import (
+    autologging_integration,
+    safe_patch,
+)
+from mlflow.utils.databricks_utils import is_in_databricks_runtime
+
+FLAVOR_NAME = "litellm"
+
+
+@experimental
+def autolog(
+    log_traces: bool = True,
+    disable: bool = False,
+    silent: bool = False,
+):
+    """
+    Enables (or disables) and configures autologging from LiteLLM to MLflow. Currently, MLflow
+    only supports autologging for tracing.
+
+    Args:
+        log_traces: If ``True``, traces are logged for LiteLLM calls. If ``False``,
+            no traces are collected during inference. Default to ``True``.
+        disable: If ``True``, disables the LiteLLM autologging integration. If ``False``,
+            enables the LiteLLM autologging integration.
+        silent: If ``True``, suppress all event logs and warnings from MLflow during LiteLLM
+            autologging. If ``False``, show all events and warnings.
+    """
+    import litellm
+
+    # This needs to be called before doing any safe-patching (otherwise safe-patch will be no-op).
+    _autolog(log_traces=log_traces, disable=disable, silent=silent)
+
+    if log_traces and not disable:
+        litellm.success_callback = _append_mlflow_callbacks(litellm.success_callback)
+        litellm.failure_callback = _append_mlflow_callbacks(litellm.failure_callback)
+
+        if is_in_databricks_runtime():
+            # Patch main APIs e.g. completion to inject custom handling for threading.
+            # By default, those API will start a new thread when calling log_success_event()
+            # handler of the logging callbacks and never wait for it to finish. This is
+            # problematic in Databricks notebook, because the inline trace UI display
+            # assumes that the trace is generated synchronously. If the trace is generated
+            # asynchronously, it will be displayed in different later cells.
+            # To workaround this issue, we monkey-patch these APIs to wait for the logging
+            # threads to finish before returning the result.
+            # This is not required for OSS environment where we don't show inline trace UI.
+            for func in [
+                "completion",
+                "embedding",
+                "text_completion",
+                "image_generation",
+                "transcription",
+                "speech",
+            ]:
+                _patch_theading_in_function(litellm, func)
+
+            # For streaming case, we need to patch the iterator because traces are generated
+            # when consuming the generator, not when calling the main APIs.
+            _patch_theading_in_function(litellm.utils.CustomStreamWrapper, "__next__")
+
+            # NB: We don't need to patch async function because Databricks notebook waits
+            # for the async task to finish before finishing the cell.
+    else:
+        litellm.success_callback = _remove_mlflow_callbacks(litellm.success_callback)
```

Comment:
Should we remove mlflow callbacks if it's added by the user?

Reply:
I think this question matches your comment on `_autolog` about `This prevents us from removing our callback and patching when autologging is turned off.` But I though autologging means, if the model doesn't contain any callback then we inject them, if it already contains we do nothing (same for langchain)

Reply:
> Should we remove mlflow callbacks if it's added by the user?

Technically not, but in practice, I wonder how often users will inject MLflow callback by themselves - imo it's near zero. We could have a global state to determine if the callback is injected by MLflow or not, but I double it gives more benefit than its complexity🤔

Reply:
Oh the callback is global in liteLLM? I thought it's defined within each model call.... Emm if so then seems we have to remove mlflow callback when disable=True..

Reply:
Right, it is global callback similar to LlamaIndex, unlike LangChain